client/src/main/scala/io/mediachain/client/CanonicalFolder.scala

```diff
@@ -0,0 +1,97 @@
+package io.mediachain.client
+
+import io.mediachain.protocol.Datastore._
+
+class CanonicalFolder(datastore: Datastore) {
+  import io.mediachain.util.cbor.CborAST.CValue
+
+  /**
+    * Tries to extend the list of `cells` by following the `chain`
+    * reference of the cell at the head of the list.  If successful,
+    * prepends the new cell to the head of the list and repeats until
+    * a nil `chain` reference is found, or the referenced cell can't
+    * be retrieved from the datastore.
+    *
+    * @param cells - a list of `ChainCell`s to extend
+    * @tparam T - Should be either EntityChainCell or ArtefactChainCell -
+    *           using a concrete subtype may fail to find all cells.
+    * @return - all the `ChainCell`s that can be found by chasing
+    *         `chain` references from the head of the input list
+    */
+  def extendChain[T <: ChainCell](cells: List[T]): List[T] = {
```

Comment:
I would also maybe break up `cells` into a `cells` accumulator and `nextCellOpt` param, rather than passing it together